### PR TITLE
Feat/pca based tac extractions

### PR DESCRIPTION
**Pull request by FurqanDar**
Added utility functions to allow for temporal PCA-based analyses of 4D-PET images with a given mask. 
 - Computing the temporal PCA over the voxels given a mask
 - Extracting the temporal PCA components (eigenvectors)
 - Extracting the temporal PCA projections
 - Filtering voxels given some thresholding quantiles over given PCA projection components and generating TACs from the filtered voxels
 - Aggregate function to generate the filtered voxels TACs using only image paths

### CODE REVIEW

File: docs/conf.py [hunk 1]
```diff
@@ -3,10 +3,10 @@
 import sys
 sys.path.insert(0, os.path.abspath('../petpal'))
 
-project = 'PET Processing Module'
+project = 'PETPAL (Positron Emission Tomography Analysis Library)'
```
Comment (kenanoestreich) [inline docs/conf.py:6]: **Processing** and Analysis Library, right? Processing's the P.

File: petpal/utils/data_driven_image_analyses.py [hunk 2]
```diff
@@ -0,0 +1,566 @@
...
+          of information.
+        - Whitening normalizes the data, which can be important if further statistical operations
+          (e.g., clustering) are planned.
+
+    See Also:
+        - :func:`extract_roi_voxel_tacs_from_image_using_mask<petpal.preproc.image_operations_4d.extract_roi_voxel_tacs_from_image_using_mask>`: Extracts voxel-level TACs for regions of interest
+          in the input image, used in TAC calculations here.
+        - :class:`sklearn.decomposition.PCA`: Underlying PCA implementation used in this function.
+    """
+    assert len(input_image.shape) == 4, "Input image must be 4D."
```
Comment (kenanoestreich) [inline petpal/utils/data_driven_image_analyses.py:78]: I think these two checks are also at the beginning of the next function, fwiw.

File: petpal/utils/data_driven_image_analyses.py
Comment (kenanoestreich) [inline petpal/utils/data_driven_image_analyses.py:1]: Looks good; I don't yet understand the difference between the two big functions named extract* and generate*.

File: docs/conf.py (hunk 1 above)
Comment (FurqanDar) [reply docs/conf.py:6]: LOL yes! Thanks for the catch!

File: petpal/utils/data_driven_image_analyses.py (hunk 2 above)
Comment (FurqanDar) [reply petpal/utils/data_driven_image_analyses.py:78]: Yeah this a sign of some overall encapsulation that could happen. For now, I wanted the called function to raise as early as possible

File: petpal/utils/data_driven_image_analyses.py
Comment (FurqanDar) [reply petpal/utils/data_driven_image_analyses.py:1]: Extract does the actual computation, whereas the generate function takes the image _paths_ and so can read the metadata and generate real TACs

File: petpal/utils/data_driven_image_analyses.py [hunk 3]
```diff
@@ -0,0 +1,566 @@
...
+
+
+def generate_temporal_pca_quantile_threshold_tacs_of_image_over_mask(input_image_path: str,
+                                                                     mask_image_path: str,
+                                                                     output_arrays_path: str | None,
+                                                                     num_components: int,
+                                                                     threshold_components: list | None,
+                                                                     quantiles: list | None,
+                                                                     direction: str = '>',
+                                                                     **sklearn_pca_kwargs
```
Comment (noahg-neuroimage) [inline petpal/utils/data_driven_image_analyses.py:383]: This function takes a lot of inputs. Would this be better implemented as a class?

File: petpal/utils/data_driven_image_analyses.py [hunk 4]
```diff
@@ -0,0 +1,566 @@
...
+    tacs_ext_func = extract_temporal_pca_quantile_thresholded_tacs_of_image_using_mask
+    tacs_mean, tacs_std = tacs_ext_func(input_image=input_image,
+                                        mask_image=mask_image,
+                                        num_components=num_components,
+                                        threshold_components=threshold_components,
+                                        quantiles=quantiles,
+                                        direction=direction,
+                                        **sklearn_pca_kwargs
+                                        )
+
```
Comment (noahg-neuroimage) [inline petpal/utils/data_driven_image_analyses.py:490]: This function is doing a lot of small tasks with logic that may be useful elsewhere. Recommend implementing as a class with separate methods.

File: petpal/utils/data_driven_image_analyses.py [hunk 5]
```diff
@@ -0,0 +1,566 @@
...
+        np.ndarray:
+            Array containing the aggregated multiple-tacs array. The first column is the time,
+            the rest of the column pairs correspond to the threshold component and the quantile.
+    """
+    assert tacs_mean.shape == tacs_std.shape, (f"The TACs mean and stderrs must have the same shape."
+                                               f" mean:{tacs_mean.shape}, std:{tacs_std.shape}")
+    num_components, num_quantiles, num_frames = tacs_mean.shape
+    num_tacs = num_components * num_quantiles
+    num_columns = 2 * num_tacs + 1
+    out_tacs_array = np.zeros(shape=(num_columns, num_frames), dtype=float)
```
Comment (noahg-neuroimage) [inline petpal/utils/data_driven_image_analyses.py:560]: Is this logic useful elsewhere as an independent function?

**Review by noahg-neuroimage — APPROVED** [review]
👍 

Good:
- Exceptional user inputs handled
- Passes GitHub actions
- New methods can be tested

Could be improved:
- New PCA functions have identical inputs and simple logic. These may be simplified by implementing as class methods to ensure DRY and ETC.
- (e.g. a variable name change or image handling logic changes could cause unexpected issues)
- Strongly encourage implementing "generate_temporal_pca_quantile_threshold_tacs_of_image_over_mask" as a class for longevity.

**Review by kenanoestreich — APPROVED** [review]
Refactor to a class in a future PR
